### PR TITLE
Add stricter parameter types for Object.values and Object.entries

### LINKED ISSUE #20550: Object.values should not accept undefined or null
**TypeScript Version:**  2.7.0-dev.201xxxxx
Reproduced locally with `2.5.3` but it seems to still be the case in master. https://github.com/Microsoft/TypeScript/blob/master/src/lib/es2017.object.d.ts

**Code**

```ts
const foo = Object.values(undefined);
```

**Expected behaviour:**
This should be a compile error, like `Object.keys(undefined)` is.

The type signature for `Object.keys` is
```ts
keys(o: {}): string[];
```
but for `Object.values` and `Object.entities` it is:

```ts
values<T>(o: { [s: string]: T }): T[];
values(o: any): any[];

entries<T>(o: { [s: string]: T }): [string, T][];
entries(o: any): [string, any][];
```

Ideally it would be nice to have specialisations for if `o: string` as well, but at minimum it would be nice to at least replace `o: any` with `o: {}` to catch the above error.

**Actual behaviour:**
This compiles correctly and gives a runtime error.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #20550

Added ` | string | number | boolean | symbol` to avoid falling into the second overload unnecessarily; `Object.values` on a primitive returns `[]` of these so it's effectively a `never[]` in that case (though we infer `{}[]`).
I'm also not too happy about returning `any` from builtin functions. `--noImplicitAny` won't help you there. The only case I found where the first overload won't match any more is `Object.values(() => {})`, in which case a return of `{}[]` is probably better since a function's unlikely to have the properties you intend to get.

### CODE REVIEW

**Comment by weswigham** [comment]
@andy-ms We could just define values as `values<T>(o: T): T[keyof T][];` (and similarly for entries). These definitions are pre-index types, and we can just do better now, I think.

**Comment by Kovensky** [comment]
@weswigham the problem there is the unsoundness because of structural non-sealed types

```ts
const a = { a: 'a string', b: 123 }
const b: Pick<typeof a, 'a'> = a
const values = Object.values(b) // string[]; but actually has a number inside too
```

**Comment by andy-ms** [comment]
@weswigham What would you think of that signature with `T extends object`? The function will technically work at runtime if you pass in a string, it's not really useful unless you're passing in an `object`.

**Comment by weswigham** [comment]
@Kovensky any time you alias a type you lose information w.r.t. the index operator. But the thing is, I'm pretty sure that definition (at least for `T extends object`, like @andy-ms suggests), is more practically useful, since you'll get the types for all the fields we _do_ know about (instead of the unsafe `any`). Honestly, your complaint is just a function of how the `keyof` type operator works; but the more tightly your types match your behavior, the more accurate it is. You use it on any type you've aliased (thus discarded information) and you'll get results for the alias, not the original type.

File: src/lib/es2017.object.d.ts [hunk 1]
```diff
@@ -3,25 +3,14 @@ interface ObjectConstructor {
      * Returns an array of values of the enumerable properties of an object
      * @param o Object that contains the properties and methods. This can be an object that you created or an existing Document Object Model (DOM) object.
      */
-    values<T>(o: { [s: string]: T } | { [n: number]: T }): T[];
+    values<T extends object>(o: T): T[keyof T][];
```
Comment (mhegazy) [inline src/lib/es2017.object.d.ts]: do not use `keyof` here. same reasoning as https://github.com/Microsoft/TypeScript/issues/13989#issuecomment-279622620

File: src/lib/es2017.object.d.ts [hunk 2]
```diff
@@ -3,25 +3,14 @@ interface ObjectConstructor {
...
-     * @param o Object that contains the properties and methods. This can be an object that you created or an existing Document Object Model (DOM) object.
-     */
-    entries<T>(o: { [s: string]: T } | { [n: number]: T }): [string, T][];
 
     /**
      * Returns an array of key/values of the enumerable properties of an object
      * @param o Object that contains the properties and methods. This can be an object that you created or an existing Document Object Model (DOM) object.
      */
-    entries(o: any): [string, any][];
+    entries<T extends object>(o: T): [string, T[keyof T]][];
```
Comment (mhegazy) [inline src/lib/es2017.object.d.ts]: same here

**Comment by andy-ms** [comment]
@mhegazy @weswigham The original goal was to not accept `undefined` and `null` here. How about making this PR just chane `values(o: any): any[];` to `values(o: {}): any[]` (and the same for `entries`) and leaving the rest for later?

**Comment by mhegazy** [comment]
> The original goal was to not accept `undefined` and `null` here. How about making this PR just change `values(o: any): any[];` to `values(o: {}): any[]` (and the same for `entries`) and leaving the rest for later?
  
that would work.

**Comment by andy-ms** [comment]
Waiting for #21075, then will merge from master

**Comment by aboyton** [comment]
Thanks. Really loving the TypeScript project and how quickly and thoroughly you fix any reported bugs.

**Pull request merged**